src/circuit_builder.rs

```diff
@@ -128,56 +81,47 @@ where
             verifier
                 .verifier_key
                 .expect("Unexpected error. Missing VerifierKey in compilation"),
+            pi_pos,
         ))
     }
 
-    /// /// Return a mutable reference to the Public Inputs storage of the circuit.
-    fn get_mut_pi_positions(&mut self) -> &mut Vec<PublicInput>;
-
-    /// Return a reference to the Public Inputs storage of the circuit.
-    fn get_pi_positions(&self) -> &Vec<PublicInput>;
-
     /// Build PI vector for Proof verifications.
-    fn build_pi(&self, pub_inputs: &[PublicInput]) -> Vec<BlsScalar> {
-        let mut pi = vec![BlsScalar::zero(); self.get_trim_size()];
-        pub_inputs
+    fn build_pi(
+        &self,
+        pub_input_values: &[PublicInputValue],
+        pub_input_pos: &PublicInputPositions,
```

Comment:
As mentioned in the comment related to `PublicInputPosition`:
```suggestion
        pub_input_positions: &[usize],
```

Nit: be consistent. You have `pub_input_pos` in `build_pi` but `pub_input_positions` in `gen_proof`: choose one.